hail_batch_dev.md

```diff
@@ -249,3 +249,62 @@ When users submit a batch, they specify a billing project which will be charged
 Billing projects also determine the visibility of batches. If two users use the same billing project, they can see each other's batches submitted under that billing project.
 
 For each of our [datasets](storage_policies), we have a dedicated Hail billing project. Whenever someone gets added to the dataset's permission group, they should also be added to the corresponding Hail Batch billing project.
+
+## Updating TLS (HTTPS) certificates
+
+At the moment, this just covers the Google Cloud deployment.
+
+1. Start the [`hail-setup` VM](https://console.cloud.google.com/compute/instancesDetail/zones/australia-southeast1-b/instances/hail-setup?project=hail-295901).
+1. Connect to the VM using SSH:
+
+   ```bash
+   gcloud --project=hail-295901 compute ssh hail-setup
+   ```
+
+1. Once connected, clone the GitHub repository (if not already done):
+
+   ```bash
+   git clone https://github.com/populationgenomics/hail.git
+   ```
+
+1. Make sure you're up-to-date on the `main` branch.
+
+   ```bash
+   cd hail
+   git switch main
+   git pull
+   ```
+
+1. Make sure the Google Cloud credentials have been set up:
+
+   ```bash
+   gcloud config set project hail-295901
+   gcloud config set compute/zone australia-southeast1-b
+   gcloud auth configure-docker australia-southeast1-docker.pkg.dev
+   gcloud container clusters get-credentials vdc
+   ```
+
+1. Now regenerate the certificates, which might take a couple of minutes:
+
+   ```bash
+   cd letscencrypt
+   make run NAMESPACE=default
+   ```
+
+1. Once the above command  has finished successfully, it's time to [restart the Hail services](https://github.com/populationgenomics/hail/blob/main/dev-docs/tls-cookbook.md#regenerate-all-the-certificates) to pick up the new certificates. This will cause a very short downtime, but won't interrupt any running batches:
+
+   ```bash
+   export HAIL=$HOME/hail
+
+   SERVICES_TO_RESTART=$(python3 -c 'import os
+   import yaml
+   hail_dir = os.getenv("HAIL")
+   x = yaml.safe_load(open(f"{hail_dir}/tls/config.yaml"))["principals"]
+   print(",".join(x["name"] for x in x))')
+
+   kubectl delete pods -l "app in ($SERVICES_TO_RESTART)"
+   ```
+
+1. [Verify in a browser](https://batch.hail.populationgenomics.org.au/) that the expiration date for the certificate has been extended.
+
+1. Don't forget to stop the [`hail-setup` VM](https://console.cloud.google.com/compute/instancesDetail/zones/australia-southeast1-b/instances/hail-setup?project=hail-295901). Please don't delete it!
```

Comment:
@lgruen, can you create a machine image for the hail-setup VM so if it does get deleted we don't have to set it up again?

Reply:
There's actually nothing special going on there (I think). It'll be interesting to see whether @violetbrina will run into any issues, as she might not have used the VM before, whereas I've got a bunch of settings in my profile.

The main reason to use a VM in the first place is to have a much faster Docker build experience and better network bandwidth for pulling / pushing images.

Reply:
Ok I've successfully gone through the whole process. The only (I suspect one time thing) I had to do was install the gke-gcloud-auth-plugin component on the gcloud cli to get everything to work.

My suggestion on the docs though is *how* exactly to check the expiration date for the certificate. That is unclear and was the one thing I couldn't do.

Reply:
Thanks! I've updated the instructions to also mention installing the gke-gcloud-auth-plugin and provided a command for the TLS certificate expiration date check.

If you want to do that in the browser, you can click the lock icon:
![CleanShot 2023-05-09 at 08 20 58@2x](https://user-images.githubusercontent.com/2876012/236950541-d0868557-7be6-41be-9b2c-b7b48e248657.png)

Interestingly, your update did *not* extend the expiration date:

```bash
> echo | openssl s_client -servername batch.hail.populationgenomics.org.au -connect batch.hail.populationgenomics.org.au:443 2>/dev/null | openssl x509 -noout -dates

notBefore=Feb 28 21:17:11 2023 GMT
notAfter=May 29 21:17:10 2023 GMT
```

Are you sure the `letsencrypt` step (6) finished successfully?